src/main/java/io/quarkus/security/PermissionChecker.java

```diff
@@ -0,0 +1,66 @@
+package io.quarkus.security;
+
+import java.lang.annotation.ElementType;
+import java.lang.annotation.Retention;
+import java.lang.annotation.RetentionPolicy;
+import java.lang.annotation.Target;
+
+/**
+ * Used to annotate CDI bean methods that check if a user holds a permission specified by the {@link #value()}.
+ * Quarkus Security will augment every {@link io.quarkus.security.identity.SecurityIdentity} with this permission checker.
+ * Such a permission checker grants access to methods secured with the {@link PermissionsAllowed} security annotation.
+ * Following example shows how it works:
+ * <pre>
+ * {@code
+ * &#64;Path("hello")
+ * public class HelloResource {
+ *
+ *     &#64;PermissionsAllowed("speak")
+ *     &#64;GET
+ *     public String sayHello() {
+ *         return "Hello World!";
+ *     }
+ *
+ *     &#64;PermissionChecker("speak")
+ *     public boolean canSpeak(SecurityIdentity identity) {
+ *         return "speaker".equals(identity.getPrincipal().getName());
+ *     }
+ * }
+ * }
+ * </pre>
+ * The permission checker methods can include any of secured method parameters (matched by name).
+ * Consider the following secured method:
+ * <pre>
+ * {@code
+ * &#64;PermissionsAllowed("update")
+ * public String updateString(String a, String b, String c, String d) {
+ *     ...
+ * }
+ * }
+ * </pre>
+ * The permission checker that grants access to the {@code updateString} method can inject
+ * any arguments it requires and optionally even {@link io.quarkus.security.identity.SecurityIdentity}:
+ * <pre>
+ * {@code
+ * &#64;PermissionChecker("update")
+ * public boolean canUpdate(String c, String a, SecurityIdentity identity) {
+ *     ...
+ * }
+ * }
+ * </pre>
+ * The permission checker method parameters are matched with the secured method parameters in exactly same fashion
+ * as are constructor parameters of a custom permission. Please see {@link PermissionsAllowed#params()} for more information.
+ */
+@Target(ElementType.METHOD)
+@Retention(RetentionPolicy.RUNTIME)
+public @interface PermissionChecker {
```

Comment:
@michalvavrik I think the PR description says this is how it would work but let me confirm.

It would be great if one could put it on any random bean, Quarkus would access a bean using Arc and call the checker method.

The example above with `HelloResource` is nice but I'm not sure if this is what users would want to do, to go via the `PermissionAllowed` indirection just so that the same endpoint method's can be called via reflection. This is a more complex version of what Steph has suggested where you can just type the security check code in the endpoint...

IMHO
```
public class HelloResource {

 @PermissionsAllowed("owner")
 @GET
 public String sayHello() {
         return "Hello World!";
 } 
```

```
public class PermissionDBChecker extends PanacheEntity {
      @PermissionChecker("owner")
      public boolean isOwner(SecurityIdentity identity, String orderId) {
          return ...
     }
 }
```

or similar would be more useful and closer to the ReBac options as well. Is it what you plan to support ?

Reply:
This will work. I have already integrated `QuarkusPermission` in my local branch and all that needs to be done is to implement `@PermissionChecker` which is mostly about build-time logic inside Quarkus Security deployment and about accessing Arc container.

There can be inheritance scenarios that could result in ambiguous resolution, like if `@PermissionChecker` is specified on a class that is implemented by 2 beans. I'll add validation to prevent it.

Reply:
wait a sec, where `String orderId` comes from?

Reply:
> indirection just so that the same endpoint method's can be called via reflection

FYI no reflection apart of class initialization is going to happen; as long as we generate classes in the same package we can access non-private methods I think; anyway I didn't add any promises on visibility to API to avoid establishing any contract there. Hopefully it's minor detail.

Reply:
One argument I can provide is that you just need plain java class, no CDI bean. And if I need a class rather than bean with method, it depends on my preference.

Reply:
@michalvavrik Sorry for the delay, sounds good, just re

>wait a sec, where String orderId comes from?

From a method parameter to which `@PermissionAllowed` is attached ?

Reply:
Yeah, I see now what you meant, I'm not sure it is really that great if we can't pass extra parameters, what would be a difference from just checking security identity in the augmentor and throwing exception if something does not fit, with the augmentor having access to various beans ?

Reply:
>  I'm not sure it is really that great if we can't pass extra parameters

You can pass the arguments, please checkout `QuarkusPermissionMethodArgsOnlyTest` in the https://github.com/quarkusio/quarkus/commit/4d50853fb2e82e6472e45a4f6ea97bb9dbdb6f85

> what would be a difference from just checking security identity in the augmentor and throwing exception if something does not fit, with the augmentor having access to various beans ?

augmentor is ok, though I must honestly say I never thought that using augmentor for authorization is good. I would say augmentor is for augmentation.

Reply:
Michal, @michalvavrik 

>You can pass the arguments, please checkout QuarkusPermissionMethodArgsOnlyTest

So what happens if I have

```
class HelloResource {
 @GET
 @PermissionAllowed("owner")
 String get(@QueryParam("repositoryid") String repoId) {}
}

class Bean {
   @PermissionChecker
   boolean isOwner(String repoId) {...}
}
```

How does `QuarkusPermission` come into the picture, do you have Quarkus instance subclass generated, with constructor accepting `String repoId`, and then its `isGranted` method calling `bean.isOwner(repoId)` ?

Reply:
This might change as I have only implemented QuarkusPermission, not the checker itself, but this is my plan ATM:

GENERATED CLASS (as isOwner is package-private, must be in the same package as the Bean):
```
// name shall be different
public class BeanIsOwnerPermission extends QuarkusPermission {

   // maybe it will be String, will see
   private final Object one;
   private final InstanceHandle<SecurityIdentity> instanceHandle;
   
  public BeanIsOwnerPermission(Object one) {
        this.one = one;
        // do some validation whether been is resolvable etc.
        this.instanceHandle = Arc.container().instance(Bean.class);
  }

    @Override
    public boolean isGranted(SecurityIdentity securityIdentity) {
        // bean resolution will be validated before hand
        // I hope to find a way how to keep instance handle on th
        return instanceHandle.get().isOwner((String)one);
    }
}
```

We can optimize this with method handles as we did for bean params in the future. But not in the first version. I don't think it's a big deal there is extra class TBH.